torchtune/utils/_generation.py

```diff
@@ -4,141 +4,115 @@
-            the top_k probabilities. Default is None
-        stop_tokens (Optional[Set[int]]): If specified, generation is stopped when any of these
-            tokens are generated. Default: None
-        custom_generate_next_token (Optional[Callable]): If specified, we'll use the custom
-            generate_next_token function (e.g. compiled function) when generating the tokens,
-            otherwise we'll use the default `geenrate_next_token` function. Default is None
+        prompt (torch.Tensor): tensor with the token IDs associated with the given prompt
+        max_generated_tokens (int): number of tokens to be generated
+        temperature (float): value to scale the predicted logits by
+        top_k (Optional[int]): If specified, we prune the sampling to only token ids within the top_k probabilities
+        stop_tokens (Optional[List[int]]): If specified, generation is stopped when any of these tokens are generated
+        custom_generate_next_token (Optional[Callable]): If specified, we'll use the custom generate_next_token function
 
     Returns:
         List: list of generated tokens
-
-    Raises:
-        ValueError: if max_seq_len supported by the model is smaller than the number of tokens
-            requested
     """
-    prompt_length = prompt.size(0)
-
-    if model.max_seq_len < (prompt_length + max_generated_tokens) - 1:
-        raise ValueError(
-            f"Models maximum seq length {model.max_seq_len} should be >= "
-            f"{(prompt_length + max_generated_tokens)} - 1"
-        )
+    prompt = prompt.view(1, -1) if prompt.ndim == 1 else prompt
+    stop_tokens = (
+        torch.tensor(stop_tokens, device=prompt.device) if stop_tokens else None
+    )
+    bsz, seq_length = prompt.size()
+    generated_tokens = prompt.clone()
+    stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
+    stop_token_mask = torch.ones(
+        (bsz, seq_length + 1), dtype=torch.int32, device=prompt.device
+    )
 
     if custom_generate_next_token is None:
         custom_generate_next_token = generate_next_token
 
-    # generated_tokens is a list of tensors where each tensor contains tokens
-    # needed for the output
-    generated_tokens = [prompt]
-
-    # generate the first token by conditioning on the input prompt
+    # generate the first token conditioned on the prompt
     token = generate_next_token(
-        model=model,
-        input_pos=torch.arange(0, prompt_length, device=prompt.device),
-        # convert the input into [B, S] shape as expected by the model
-        x=prompt.view(1, -1),
+        model,
+        input_pos=torch.arange(0, seq_length, device=prompt.device),
+        x=prompt,
         temperature=temperature,
         top_k=top_k,
-    ).clone()
+    )
+    generated_tokens = torch.cat([generated_tokens, token], dim=-1)
 
-    generated_tokens.append(token)
+    # stop early if we reach a stop token in every seq
+    if stop_tokens is not None:
```

Comment:
I guess it makes sense that there is a chance that the prompt itself generates a stop token. But I think this can be converted into a simple utility within this file since you need it in a couple of places